client/data/emails/use-remove-email-forward-mutation.ts

```diff
@@ -27,24 +38,105 @@ export function useRemoveEmailForwardMutation(
 	const domainsQueryKey = getEmailDomainsQueryKey( selectedSiteId );
 
 	const suppliedOnSettled = mutationOptions.onSettled;
+	const suppliedOnMutate = mutationOptions.onMutate;
+	const suppliedOnError = mutationOptions.onError;
+	const suppliedOnSuccess = mutationOptions.onSuccess;
+
+	mutationOptions.mutationKey = MUTATION_KEY;
 
-	mutationOptions.onSettled = async ( data, error, variables, context ) => {
+	mutationOptions.onSettled = ( data, error, variables, context ) => {
 		suppliedOnSettled?.( data, error, variables, context );
 
-		await Promise.all( [
-			queryClient.invalidateQueries( emailDomainsQueryKey ),
-			queryClient.invalidateQueries( domainsQueryKey ),
-		] );
+		queryClient.invalidateQueries( emailDomainsQueryKey );
+		queryClient.invalidateQueries( domainsQueryKey );
 	};
 
-	return useMutation(
-		() =>
-			wp.req.get( {
-				path: `/domains/${ encodeURIComponent( domainName ) }/email/${ encodeURIComponent(
-					mailboxName
-				) }/delete`,
-				method: 'POST',
+	mutationOptions.onMutate = async ( emailForward ) => {
+		suppliedOnMutate?.( emailForward );
+
+		await queryClient.cancelQueries( emailDomainsQueryKey );
+
+		// Optimistically remove email forward from `useGetEmailAccountsQuery` data
+		const previousEmailAccountsQueryData = queryClient.getQueryData< any >( emailDomainsQueryKey );
+
+		const emailForwards = previousEmailAccountsQueryData.accounts?.[ 0 ]?.emails;
+
+		if ( emailForwards ) {
+			queryClient.setQueryData( emailDomainsQueryKey, {
+				...previousEmailAccountsQueryData,
+				accounts: [
+					{
+						...previousEmailAccountsQueryData.accounts[ 0 ],
+						emails: emailForwards.filter(
+							( forward: Mailbox ) => forward.mailbox !== emailForward.mailbox
+						),
+					},
+				],
+			} );
+		}
+
+		return {
+			[ JSON.stringify( emailDomainsQueryKey ) ]: previousEmailAccountsQueryData,
+		};
+	};
+
+	mutationOptions.onSuccess = ( data, emailForward, context ) => {
+		suppliedOnSuccess?.( data, emailForward, context );
+
+		const emailAddress = getEmailAddress( emailForward );
+
+		const successMessage = successNotice(
+			translate( '{{strong}}%(emailAddress)s{{/strong}} has been removed from your account', {
+				comment:
+					'%(emailAddress)s is the receiver email address for the email forward being deleted',
+				args: { emailAddress },
+				components: {
+					strong: createElement( 'strong' ),
+				},
 			} ),
+			{ duration: 7000 }
+		);
+
+		dispatch( successMessage );
+	};
+
+	mutationOptions.onError = ( error, emailForward, context ) => {
+		suppliedOnError?.( error, emailForward, context );
+
+		if ( context ) {
+			queryClient.setQueryData(
+				emailDomainsQueryKey,
+				context[ JSON.stringify( emailDomainsQueryKey ) ]
+			);
+		}
+
+		const emailAddress = getEmailAddress( emailForward );
+
+		const errorMessage = errorNotice(
+			translate(
+				'There was an error removing {{strong}}%(emailAddress)s{{/strong}} from your account',
```

Comment:
Note we end all error messages with a dot, unless there is a specific reason not to do so:

```suggestion
				'There was an error removing {{strong}}%(emailAddress)s{{/strong}} from your account.',
```

Reply:
Also could we keep all those error messages consistent?

> Failed to remove email forward. Please try again or contact support.

I like the idea of mentioning the address like you did though:

> Failed to remove your email forward for {{strong}}%(emailAddress)s{{/strong}}. Please try again or contact support.

Note I don't mind revisiting those error messages, I just want to make sure that they follow the same patterns.

Reply:
I went over the notice messages in the add/remove mutations and standardized the phrasing a little bit. I also made sure that the address of the email forward is mentioned in all of them.